pkg/k8s/tcpportmapping.go

```diff
@@ -0,0 +1,177 @@
+package k8s
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+	"sync"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/util/retry"
+)
+
+// TCPPortMapping is a TCPPortMapper backed by a Kubernetes ConfigMap.
+type TCPPortMapping struct {
+	mu    sync.RWMutex
+	table map[int32]*ServiceWithPort
+
+	minPort int32
+	maxPort int32
+
+	client          kubernetes.Interface
+	cfgMapNamespace string
+	cfgMapName      string
+}
+
+// NewTCPPortMapping creates a new TCPPortMapping instance.
+func NewTCPPortMapping(client kubernetes.Interface, cfgMapNamespace, cfgMapName string, minPort, maxPort int32) (*TCPPortMapping, error) {
+	m := &TCPPortMapping{
+		minPort:         minPort,
+		maxPort:         maxPort,
+		table:           make(map[int32]*ServiceWithPort),
+		client:          client,
+		cfgMapNamespace: cfgMapNamespace,
+		cfgMapName:      cfgMapName,
+	}
+
+	if err := m.loadState(); err != nil {
+		return nil, err
+	}
+
+	return m, nil
+}
+
+// Find searches for the port which is associated with the given ServiceWithPort.
+func (m *TCPPortMapping) Find(svc ServiceWithPort) (int32, bool) {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+
+	for port, v := range m.table {
+		if v.Name == svc.Name && v.Namespace == svc.Namespace && v.Port == svc.Port {
+			return port, true
+		}
+	}
+
+	return 0, false
+}
+
+// Get returns the ServiceWithPort associated to the given port.
+func (m *TCPPortMapping) Get(srcPort int32) *ServiceWithPort {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+
+	return m.table[srcPort]
+}
+
+// Add adds a new mapping between the given ServiceWithPort and the first port available in the range defined
+// within minPort and maxPort. If there's no port left, an error will be returned.
+func (m *TCPPortMapping) Add(svc *ServiceWithPort) (int32, error) {
+	for i := m.minPort; i < m.maxPort+1; i++ {
+		// Skip until an available port is found
+		if _, exists := m.table[i]; exists {
+			continue
+		}
+
+		m.mu.Lock()
+		m.table[i] = svc
+		m.mu.Unlock()
+
+		if err := m.saveState(); err != nil {
+			// If the state can't be saved, we are going to have a mismatch between the local table and the ConfigMap.
+			// By not undoing the assignment on the local table we allow the state to converge in future calls to Add,
+			// making it more robust to temporary failure.
+			return 0, fmt.Errorf("unable to save TCP port mapping: %w", err)
+		}
+
+		return i, nil
+	}
+
+	return 0, errors.New("unable to find an available port")
+}
+
+func (m *TCPPortMapping) loadState() error {
+	cfg, err := m.client.CoreV1().ConfigMaps(m.cfgMapNamespace).Get(m.cfgMapName, metav1.GetOptions{})
```

Comment:
Should use lister, not client getter.

Reply:
Yes, we can use a lister. But if we do so, we shouldn't create the lister from the shared informer `kubernetesFactory` and instead create a new informer especially for this resource. If we reuse the same informer, like we did before, here's what is going to happen:
- `controller.StartInformer`: We start all informers and wait for there cache to be filled. At this point we start receiving events on the `handler`.
- `handler.OnAdd` gets called which calls `controller.createMeshService`. This function relies on the TCPStateTable, which is not initialized yet.
So, the TCPStateTable needs to be initialized before we receive events on the informer but after startInformer since we need to informer to be populated. This is not something we can do unless creating a different informer.

We chose to not use an informer because we were doing `Get`s and not `List`s, only when initializing the state and only when `Add` gets called, which is not very often (except on startup). Since We also call `Update` in the `Add`, it means that every time we call `Add` the informer will receive the updated configmap. So, we thought, not using an informer would result in (almost) the same amount of traffic and calls would be at the same frequency. One thing that motivated us to not use Informers in the case is that we are interested by a single configmap, so the informer index would be full of unwanted objects.

Reply:
Makes sense.
We have been focusing on using listers for performance reasons, but you are right, we are only dealing with one configmap, so `Get()` should be fine.